curator-framework/src/main/java/org/apache/curator/framework/imps/NamespaceWatcher.java

```diff
@@ -75,4 +89,53 @@ else if ( curatorWatcher != null )
             }
         }
     }
+
+    // specialized equals()/hashCode() that makes this instance equal to the actual watcher
+
+    @Override
+    public boolean equals(Object o)
+    {
+        if ( this == o )
+        {
+            return true;
+        }
+        if ( o == null )
+        {
+            return false;
+        }
+
+        if ( getClass() == o.getClass() )
+        {
+            NamespaceWatcher watcher = (NamespaceWatcher)o;
+
+            if ( !unfixedPath.equals(watcher.getUnfixedPath()) )
+            {
+                return false;
+            }
+
+            //noinspection SimplifiableIfStatement
+            if ( actualWatcher != null ? !actualWatcher.equals(watcher.actualWatcher) : watcher.actualWatcher != null )
+            {
+                return false;
+            }
+            return curatorWatcher != null ? curatorWatcher.equals(watcher.curatorWatcher) : watcher.curatorWatcher == null;
+        }
+
```

Comment:
I don't think it can actually happen as my main concern is inside ZK. But, for completeness it should be added.